File: examples/cs2d/src/game/systems/BombSystem.ts
import { Player, GameState } from '../GameCore';
import { Vector2D } from '../physics/PhysicsEngine';
import { CS16AudioManager } from '../audio/CS16AudioManager';

export interface BombSite {
  id: string;
  position: Vector2D;
  radius: number;
  name: string; // 'A' or 'B'
  isActive: boolean;
}

export interface C4Bomb {
  id: string;
  position: Vector2D;
  planterId: string;
  plantTime: number;
  timer: number; // seconds until explosion
  isDefusing: boolean;
  defuserId?: string;
  defuseStartTime?: number;
  hasExploded: boolean;
  hasBeenDefused: boolean;
}

export interface DefuseAttempt {
  playerId: string;
  bombId: string;
  startTime: number;
  defuseTime: number; // How long defuse takes
  hasDefuseKit: boolean;
  isComplete: boolean;
  wasInterrupted: boolean;
}

export class BombSystem {
  private audioManager: CS16AudioManager | null = null;
  private bombSites: Map<string, BombSite> = new Map();
  private activeBombs: Map<string, C4Bomb> = new Map();
  private currentDefuse: DefuseAttempt | null = null;
  private eventCallbacks: Map<string, Function[]> = new Map();
  
  // CS 1.6 bomb mechanics constants
  private readonly BOMB_TIMER = 45; // 45 seconds
  private readonly PLANT_TIME = 3; // 3 seconds to plant
  private readonly DEFUSE_TIME_NO_KIT = 10; // 10 seconds without kit
  private readonly DEFUSE_TIME_WITH_KIT = 5; // 5 seconds with kit
  private readonly EXPLOSION_RADIUS = 500; // pixels
  private readonly EXPLOSION_DAMAGE = 500; // max damage at center
  
  constructor(audioManager?: CS16AudioManager) {
    this.audioManager = audioManager || null;
    this.initializeBombSites();
  }
  
  /**
   * Initialize bomb sites for the map
   */
  private initializeBombSites(): void {
    // Default bomb sites (would be loaded from map data)
    const bombSiteA: BombSite = {
      id: 'site_a',
      position: { x: 200, y: 200 },
      radius: 100,
      name: 'A',
      isActive: true
    };
    
    const bombSiteB: BombSite = {
      id: 'site_b',
      position: { x: 800, y: 600 },
      radius: 100,
      name: 'B',
      isActive: true
    };
    
    this.bombSites.set('site_a', bombSiteA);
    this.bombSites.set('site_b', bombSiteB);
  }
  
  /**
   * Update bomb system each frame
   */
  update(deltaTime: number): void {
    this.updateBombTimers(deltaTime);
    this.updateDefuseAttempts(deltaTime);
  }
  
  /**
   * Update bomb timers and check for explosions
   */
  private updateBombTimers(deltaTime: number): void {
    this.activeBombs.forEach((bomb) => {
      if (bomb.hasExploded || bomb.hasBeenDefused) return;
      
      bomb.timer -= deltaTime;
      
      // Play beeping sound as timer gets low
      if (bomb.timer <= 10 && bomb.timer % 1 < deltaTime) {
        this.playBombBeep(bomb);
      }
      
      // Explode bomb when timer reaches zero
      if (bomb.timer <= 0) {
        this.explodeBomb(bomb.id);
      }
    });
  }
  
  /**
   * Update defuse attempts
   */
  private updateDefuseAttempts(deltaTime: number): void {
    if (!this.currentDefuse || this.currentDefuse.isComplete) return;
    
    const elapsed = (Date.now() - this.currentDefuse.startTime) / 1000;
    
    if (elapsed >= this.currentDefuse.defuseTime) {
      this.completeDefuse();
    }
  }
  
  /**
   * Attempt to plant bomb at current position
   */
  canPlantBomb(player: Player): { canPlant: boolean; reason?: string; siteId?: string } {
    // Check if player is terrorist
    if (player.team !== 't') {
      return { canPlant: false, reason: 'Only terrorists can plant the bomb' };
    }
    
    // Check if player is alive
    if (!player.isAlive) {
      return { canPlant: false, reason: 'Player is not alive' };
    }
    
    // Check if player has C4
    if (!player.weapons.includes('c4')) {
      return { canPlant: false, reason: 'Player does not have the bomb' };
    }
    
    // Check if already a bomb planted
    if (this.activeBombs.size > 0) {
      return { canPlant: false, reason: 'Bomb is already planted' };
    }
    
    // Check if in bomb site
    const nearestSite = this.findNearestBombSite(player.position);
    if (!nearestSite) {
      return { canPlant: false, reason: 'Must be in bomb site to plant' };
    }
    
    const distance = this.calculateDistance(player.position, nearestSite.position);
    if (distance > nearestSite.radius) {
      return { canPlant: false, reason: 'Must be closer to bomb site center' };
    }
    
    return { canPlant: true, siteId: nearestSite.id };
  }
  
  /**
   * Start planting the bomb
   */
  startPlantBomb(player: Player): boolean {
    const plantCheck = this.canPlantBomb(player);
    if (!plantCheck.canPlant) {
      console.warn('Cannot plant bomb:', plantCheck.reason);
      return false;
    }
    
    // Remove C4 from player inventory
    const c4Index = player.weapons.indexOf('c4');
    if (c4Index > -1) {
      player.weapons.splice(c4Index, 1);
    }
    
    // Create bomb
    const bomb: C4Bomb = {
      id: `bomb_${Date.now()}`,
      position: { ...player.position },
      planterId: player.id,
      plantTime: Date.now(),
      timer: this.BOMB_TIMER,
      isDefusing: false,
      hasExploded: false,
      hasBeenDefused: false
    };
    
    this.activeBombs.set(bomb.id, bomb);
    
    // Award money for planting
    player.money += 800; // CS 1.6 plant reward
    
    // Play plant sound
    if (this.audioManager) {
      this.audioManager.play('radio/bombpl.wav');
      
      // Start bomb beeping sound loop
      setTimeout(() => this.startBombBeeping(bomb), 1000);
    }
    
    this.emitEvent('bomb_planted', {
      bombId: bomb.id,
      playerId: player.id,
      position: bomb.position,
      siteId: plantCheck.siteId,
      timer: bomb.timer
    });
    
    console.log('💣 Bomb planted by:', player.id, 'at:', bomb.position);
    return true;
  }
  
  /**
   * Check if player can defuse bomb
   */
  canDefuseBomb(player: Player): { canDefuse: boolean; reason?: string; bombId?: string } {
    // Check if player is CT
    if (player.team !== 'ct') {
      return { canDefuse: false, reason: 'Only CTs can defuse the bomb' };
    }
    
    // Check if player is alive
    if (!player.isAlive) {
      return { canDefuse: false, reason: 'Player is not alive' };
    }
    
    // Check if there's a bomb to defuse
    if (this.activeBombs.size === 0) {
      return { canDefuse: false, reason: 'No bomb is planted' };
    }
    
    // Find nearest bomb
    const nearestBomb = this.findNearestBomb(player.position);
    if (!nearestBomb) {
      return { canDefuse: false, reason: 'No bomb found' };
    }
    
    const distance = this.calculateDistance(player.position, nearestBomb.position);
    if (distance > 100) { // Must be within 100 units
      return { canDefuse: false, reason: 'Must be closer to the bomb' };
    }
    
    // Check if bomb has already exploded or been defused
    if (nearestBomb.hasExploded || nearestBomb.hasBeenDefused) {
      return { canDefuse: false, reason: 'Bomb cannot be defused' };
    }
    
    // Check if someone is already defusing
    if (this.currentDefuse && !this.currentDefuse.isComplete) {
      return { canDefuse: false, reason: 'Someone is already defusing' };
    }
    
    return { canDefuse: true, bombId: nearestBomb.id };
  }
  
  /**
   * Start defusing the bomb
   */
  startDefuseBomb(player: Player): boolean {
    const defuseCheck = this.canDefuseBomb(player);
    if (!defuseCheck.canDefuse || !defuseCheck.bombId) {
      console.warn('Cannot defuse bomb:', defuseCheck.reason);
      return false;
    }
    
    const bomb = this.activeBombs.get(defuseCheck.bombId);
    if (!bomb) return false;
    
    // Check if player has defuse kit
    const hasDefuseKit = player.weapons.includes('defuse_kit');
    const defuseTime = hasDefuseKit ? this.DEFUSE_TIME_WITH_KIT : this.DEFUSE_TIME_NO_KIT;
    
    // Start defuse attempt
    this.currentDefuse = {
      playerId: player.id,
      bombId: bomb.id,
      startTime: Date.now(),
      defuseTime,
      hasDefuseKit,
      isComplete: false,
      wasInterrupted: false
    };
    
    bomb.isDefusing = true;
    bomb.defuserId = player.id;
    bomb.defuseStartTime = Date.now();
    
    // Play defuse start sound
    if (this.audioManager) {
      // CS 1.6 doesn't have a specific defuse start sound, use generic
      this.audioManager.play('items/flashlight1.wav'); // Placeholder
    }
    
    this.emitEvent('defuse_started', {
      playerId: player.id,
      bombId: bomb.id,
      defuseTime,
      hasKit: hasDefuseKit,
      timeRemaining: bomb.timer
    });
    
    console.log('🛠️ Defuse started by:', player.id, 'time:', defuseTime, 'kit:', hasDefuseKit);
    return true;
  }
  
  /**
   * Stop defusing (player moved away or took damage)
   */
  stopDefuseBomb(playerId: string, reason: string = 'interrupted'): boolean {
    if (!this.currentDefuse || this.currentDefuse.playerId !== playerId) {
      return false;
    }
    
    const bomb = this.activeBombs.get(this.currentDefuse.bombId);
    if (bomb) {
      bomb.isDefusing = false;
      bomb.defuserId = undefined;
      bomb.defuseStartTime = undefined;
    }
    
    this.currentDefuse.wasInterrupted = true;
    this.currentDefuse.isComplete = true;
    
    this.emitEvent('defuse_stopped', {
      playerId,
      bombId: this.currentDefuse.bombId,
      reason
    });
    
    this.currentDefuse = null;
    
    console.log('🛠️ Defuse stopped by:', playerId, 'reason:', reason);
    return true;
  }
  
  /**
   * Complete defuse attempt
   */
  private completeDefuse(): void {
    if (!this.currentDefuse) return;
    
    const bomb = this.activeBombs.get(this.currentDefuse.bombId);
    if (!bomb) return;
    
    bomb.hasBeenDefused = true;
    bomb.isDefusing = false;
    this.currentDefuse.isComplete = true;
    
    // Award money for successful defuse
    const defuserId = this.currentDefuse.playerId;
    // In a real implementation, you'd access the player and award money
    
    // Play defuse success sound
    if (this.audioManager) {
      this.audioManager.play('radio/bombdef.wav');
    }
    
    this.emitEvent('bomb_defused', {
      playerId: defuserId,
      bombId: bomb.id,
      defuseTime: this.currentDefuse.defuseTime,
      hasKit: this.currentDefuse.hasDefuseKit,
      timeRemaining: bomb.timer
    });
    
    console.log('✅ Bomb defused by:', defuserId);
    this.currentDefuse = null;
  }
  
  /**
   * Explode the bomb
   */
  private explodeBomb(bombId: string): void {
    const bomb = this.activeBombs.get(bombId);
    if (!bomb || bomb.hasExploded) return;
    
    bomb.hasExploded = true;
    bomb.timer = 0;
    
    // Stop any defuse attempt
    if (this.currentDefuse && this.currentDefuse.bombId === bombId) {
      this.currentDefuse = null;
    }
    
    // Play explosion sound
    if (this.audioManager) {
      this.audioManager.play('weapons/c4_explode1.wav', bomb.position);
    }
    
    this.emitEvent('bomb_exploded', {
      bombId: bomb.id,
      position: bomb.position,
      planterId: bomb.planterId,
      explosionRadius: this.EXPLOSION_RADIUS,
      maxDamage: this.EXPLOSION_DAMAGE
    });
    
    console.log('💥 Bomb exploded:', bombId, 'at:', bomb.position);
  }
  
  /**
   * Calculate damage from bomb explosion for a player
   */
  calculateExplosionDamage(bombPosition: Vector2D, playerPosition: Vector2D): number {
    const distance = this.calculateDistance(bombPosition, playerPosition);
    
    if (distance >= this.EXPLOSION_RADIUS) {
      return 0; // Outside explosion radius
    }
    
    // Linear damage falloff
    const damageRatio = 1 - (distance / this.EXPLOSION_RADIUS);
    return Math.floor(this.EXPLOSION_DAMAGE * damageRatio);
  }
  
  /**
   * Start bomb beeping sound loop
   */
  private startBombBeeping(bomb: C4Bomb): void {
    const beep = () => {
      if (bomb.hasExploded || bomb.hasBeenDefused) return;
      
      // Beep frequency increases as timer gets lower
      let beepInterval = 1000; // 1 second default
      if (bomb.timer <= 10) beepInterval = 300; // Fast beeping
      else if (bomb.timer <= 20) beepInterval = 500; // Medium beeping
      else if (bomb.timer <= 30) beepInterval = 800; // Slow beeping
      
      this.playBombBeep(bomb);
      
      setTimeout(beep, beepInterval);
    };
    
    beep();
  }
  
  /**
   * Play bomb beep sound
   */
  private playBombBeep(bomb: C4Bomb): void {
    if (this.audioManager) {
      // CS 1.6 uses simple beep sound
      this.audioManager.play('weapons/c4_beep1.wav', bomb.position);
    }
  }
  
  /**
   * Find nearest bomb site to position
   */
  private findNearestBombSite(position: Vector2D): BombSite | null {
    let nearestSite: BombSite | null = null;
    let nearestDistance = Infinity;
    
    this.bombSites.forEach(site => {
      if (!site.isActive) return;
      
      const distance = this.calculateDistance(position, site.position);
      if (distance < nearestDistance) {
        nearestDistance = distance;
        nearestSite = site;
      }
    });
    
    return nearestSite;
  }
  
  /**
   * Find nearest active bomb
   */
  private findNearestBomb(position: Vector2D): C4Bomb | null {
    let nearestBomb: C4Bomb | null = null;
    let nearestDistance = Infinity;
    
    this.activeBombs.forEach(bomb => {
      if (bomb.hasExploded || bomb.hasBeenDefused) return;
      
      const distance = this.calculateDistance(position, bomb.position);
      if (distance < nearestDistance) {
        nearestDistance = distance;
        nearestBomb = bomb;
      }
    });
    
    return nearestBomb;
  }
  
  /**
   * Calculate distance between two points
   */
  private calculateDistance(a: Vector2D, b: Vector2D): number {
    return Math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2);
  }
  
  /**
   * Register event callback
   */
  onEvent(event: string, callback: Function): void {
    if (!this.eventCallbacks.has(event)) {
      this.eventCallbacks.set(event, []);
    }
    this.eventCallbacks.get(event)!.push(callback);
  }
  
  /**
   * Emit event to registered callbacks
   */
  private emitEvent(event: string, data: any): void {
    const callbacks = this.eventCallbacks.get(event);
    if (callbacks) {
      callbacks.forEach(callback => callback(data));
    }
  }
  
  /**
   * Get all active bombs
   */
  getActiveBombs(): C4Bomb[] {
    return Array.from(this.activeBombs.values());
  }
  
  /**
   * Get all bomb sites
   */
  getBombSites(): BombSite[] {
    return Array.from(this.bombSites.values());
  }
  
  /**
   * Get current defuse attempt
   */
  getCurrentDefuse(): DefuseAttempt | null {
    return this.currentDefuse;
  }
  
  /**
   * Check if bomb is planted
   */
  isBombPlanted(): boolean {
    return this.activeBombs.size > 0 && 
           Array.from(this.activeBombs.values()).some(bomb => 
             !bomb.hasExploded && !bomb.hasBeenDefused
           );
  }
  
  /**
   * Get defuse progress (0-1)
   */
  getDefuseProgress(): number {
    if (!this.currentDefuse || this.currentDefuse.isComplete) return 0;
    
    const elapsed = (Date.now() - this.currentDefuse.startTime) / 1000;
    return Math.min(1, elapsed / this.currentDefuse.defuseTime);
  }
  
  /**
   * Reset bomb system (for new round)
   */
  reset(): void {
    this.activeBombs.clear();
    this.currentDefuse = null;
    
    console.log('🔄 Bomb system reset for new round');
  }
  
  /**
   * Set audio manager
   */
  setAudioManager(audioManager: CS16AudioManager): void {
    this.audioManager = audioManager;
  }
}

File: examples/cs2d/game/bomb_system.rb
# frozen_string_literal: true

# CS 1.6 Bomb System
# Handles C4 bomb planting, defusing, and explosion mechanics
class BombSystem
	attr_reader :state, :planted_at_site, :planter_id, :timer, :defuser_id, :defuse_progress
	
	# CS 1.6 authentic timing values (in seconds)
	BOMB_TIMER = 45.0           # Time until explosion after planting
	PLANT_TIME = 3.0             # Time required to plant bomb
	DEFUSE_TIME = 10.0           # Time to defuse without kit
	DEFUSE_TIME_WITH_KIT = 5.0  # Time to defuse with kit
	EXPLOSION_RADIUS = 500       # Damage radius in units
	EXPLOSION_DAMAGE = 500       # Max damage at center
	BEEP_INTERVALS = [1.0, 0.8, 0.6, 0.4, 0.2, 0.1]  # Beeping acceleration
	
	# Bomb states
	STATES = {
		not_planted: 0,
		planting: 1,
		planted: 2,
		defusing: 3,
		defused: 4,
		exploded: 5
	}.freeze
	
	def initialize
		reset
	end
	
	def reset
		@state = STATES[:not_planted]
		@carrier_id = nil
		@planted_at_site = nil
		@planter_id = nil
		@defuser_id = nil
		@timer = 0
		@plant_progress = 0
		@defuse_progress = 0
		@last_beep_time = 0
		@beep_interval_index = 0
		@bomb_position = nil
	end
	
	def assign_carrier(player_id)
		return false if @state != STATES[:not_planted]
		@carrier_id = player_id
		true
	end
	
	def drop_bomb(position)
		return false unless @carrier_id
		return false if @state != STATES[:not_planted]
		
		@carrier_id = nil
		@bomb_position = position
		true
	end
	
	def pickup_bomb(player_id)
		return false if @state != STATES[:not_planted]
		return false unless @bomb_position  # Bomb must be dropped
		
		@carrier_id = player_id
		@bomb_position = nil
		true
	end
	
	def start_planting(player_id, site_name, position)
		return false unless can_plant?(player_id, site_name)
		
		@state = STATES[:planting]
		@planter_id = player_id
		@planted_at_site = site_name
		@bomb_position = position
		@plant_progress = 0
		
		true
	end
	
	def cancel_planting
		return false unless @state == STATES[:planting]
		
		@state = STATES[:not_planted]
		@plant_progress = 0
		@planter_id = nil
		@planted_at_site = nil
		
		true
	end
	
	def update_planting(delta_time, player_still_planting)
		return false unless @state == STATES[:planting]
		
		unless player_still_planting
			cancel_planting
			return false
		end
		
		@plant_progress += delta_time
		
		if @plant_progress >= PLANT_TIME
			complete_planting
			return true
		end
		
		false
	end
	
	def complete_planting
		@state = STATES[:planted]
		@timer = BOMB_TIMER
		@carrier_id = nil
		@plant_progress = 0
		@last_beep_time = 0
		@beep_interval_index = 0
		
		true
	end
	
	def start_defusing(player_id, has_kit = false)
		return false unless can_defuse?(player_id)
		
		@state = STATES[:defusing]
		@defuser_id = player_id
		@defuse_progress = 0
		@defuse_time_required = has_kit ? DEFUSE_TIME_WITH_KIT : DEFUSE_TIME
		
		true
	end
	
	def cancel_defusing
		return false unless @state == STATES[:defusing]
		
		@state = STATES[:planted]
		@defuse_progress = 0
		@defuser_id = nil
		
		true
	end
	
	def update_defusing(delta_time, player_still_defusing)
		return false unless @state == STATES[:defusing]
		
		unless player_still_defusing
			cancel_defusing
			return false
		end
		
		@defuse_progress += delta_time
		
		if @defuse_progress >= @defuse_time_required
			complete_defusing
			return true
		end
		
		false
	end
	
	def complete_defusing
		@state = STATES[:defused]
		@timer = 0
		@defuse_progress = 0
		
		true
	end
	
	def update(delta_time)
		return if @state != STATES[:planted]
		
		@timer -= delta_time
		
		if @timer <= 0
			explode
			return :exploded
		end
		
		# Check for beep
		beep_result = check_beep(delta_time)
		return beep_result if beep_result
		
		nil
	end
	
	def explode
		@state = STATES[:exploded]
		@timer = 0
		true
	end
	
	def calculate_explosion_damage(player_position)
		return 0 unless @state == STATES[:exploded]
		return 0 unless @bomb_position
		
		distance = calculate_distance(@bomb_position, player_position)
		
		return 0 if distance > EXPLOSION_RADIUS
		
		# Linear falloff
		damage_multiplier = 1.0 - (distance / EXPLOSION_RADIUS)
		(EXPLOSION_DAMAGE * damage_multiplier).to_i
	end
	
	def can_plant?(player_id, site_name)
		return false unless @carrier_id == player_id
		return false unless @state == STATES[:not_planted]
		return false unless [:a, :b].include?(site_name)
		
		true
	end
	
	def can_defuse?(player_id)
		return false unless @state == STATES[:planted]
		return false if player_id.nil?
		
		true
	end
	
	def is_carrier?(player_id)
		@carrier_id == player_id && @state == STATES[:not_planted]
	end
	
	def get_plant_progress_percentage
		return 0 unless @state == STATES[:planting]
		(@plant_progress / PLANT_TIME * 100).to_i
	end
	
	def get_defuse_progress_percentage
		return 0 unless @state == STATES[:defusing]
		(@defuse_progress / @defuse_time_required * 100).to_i
	end
	
	def get_timer_percentage
		return 0 unless @state == STATES[:planted]
		(@timer / BOMB_TIMER * 100).to_i
	end
	
	def get_beep_interval
		return nil unless @state == STATES[:planted]
		
		# Accelerate beeping as timer decreases
		time_percentage = @timer / BOMB_TIMER
		
		if time_percentage > 0.8
			BEEP_INTERVALS[0]
		elsif time_percentage > 0.6
			BEEP_INTERVALS[1]
		elsif time_percentage > 0.4
			BEEP_INTERVALS[2]
		elsif time_percentage > 0.2
			BEEP_INTERVALS[3]
		elsif time_percentage > 0.1
			BEEP_INTERVALS[4]
		else
			BEEP_INTERVALS[5]
		end
	end
	
	def get_state_info
		{
			state: STATES.key(@state),
			carrier_id: @carrier_id,
			planted_at_site: @planted_at_site,
			planter_id: @planter_id,
			defuser_id: @defuser_id,
			timer: @timer.to_i,
			timer_percentage: get_timer_percentage,
			plant_progress: get_plant_progress_percentage,
			defuse_progress: get_defuse_progress_percentage,
			bomb_position: @bomb_position,
			beep_interval: get_beep_interval
		}
	end
	
	private
	
	def check_beep(delta_time)
		current_interval = get_beep_interval
		return nil unless current_interval
		
		@last_beep_time += delta_time
		
		if @last_beep_time >= current_interval
			@last_beep_time = 0
			return :beep
		end
		
		nil
	end
	
	def calculate_distance(pos1, pos2)
		return 0 unless pos1 && pos2
		
		dx = pos1[:x] - pos2[:x]
		dy = pos1[:y] - pos2[:y]
		Math.sqrt(dx * dx + dy * dy)
	end
end

# Bomb site definitions for maps
class BombSites
	SITE_DEFINITIONS = {
		de_dust2: {
			a: {
				name: "Bombsite A",
				center: { x: 1200, y: 400 },
				radius: 150,
				plant_zones: [
					{ x: 1150, y: 350, width: 100, height: 100 }
				]
			},
			b: {
				name: "Bombsite B",
				center: { x: 400, y: 800 },
				radius: 150,
				plant_zones: [
					{ x: 350, y: 750, width: 100, height: 100 }
				]
			}
		},
		de_inferno: {
			a: {
				name: "Bombsite A",
				center: { x: 800, y: 300 },
				radius: 120,
				plant_zones: [
					{ x: 750, y: 250, width: 100, height: 100 }
				]
			},
			b: {
				name: "Bombsite B",
				center: { x: 300, y: 700 },
				radius: 120,
				plant_zones: [
					{ x: 250, y: 650, width: 100, height: 100 }
				]
			}
		}
	}.freeze
	
	def self.get_sites(map_name)
		SITE_DEFINITIONS[map_name] || SITE_DEFINITIONS[:de_dust2]
	end
	
	def self.is_in_bombsite?(position, map_name, site_name)
		sites = get_sites(map_name)
		site = sites[site_name]
		return false unless site
		
		# Check if position is within any plant zone
		site[:plant_zones].any? do |zone|
			position[:x] >= zone[:x] &&
			position[:x] <= zone[:x] + zone[:width] &&
			position[:y] >= zone[:y] &&
			position[:y] <= zone[:y] + zone[:height]
		end
	end
	
	def self.get_nearest_bombsite(position, map_name)
		sites = get_sites(map_name)
		nearest_site = nil
		nearest_distance = Float::INFINITY
		
		sites.each do |site_name, site_data|
			dx = position[:x] - site_data[:center][:x]
			dy = position[:y] - site_data[:center][:y]
			distance = Math.sqrt(dx * dx + dy * dy)
			
			if distance < nearest_distance
				nearest_distance = distance
				nearest_site = site_name
			end
		end
		
		nearest_site
	end
end